core/src/main/scala/org/labrad/data/Data.scala

```diff
@@ -1208,12 +1232,72 @@ object Arr {
   def apply(elems: Seq[Data]): Data = apply(elems.toArray)
   def apply(elem: Data, elems: Data*): Data = apply(elem +: elems)
 
-  def apply(a: Array[Boolean]): Data = make[Boolean](a, TBool)
-  def apply(a: Array[Int]): Data = make[Int](a, TInt)
-  def apply(a: Array[Long]): Data = make[Long](a, TUInt)
+  def apply(a: Array[Boolean]): Data = {
+    val elem = TBool
+    val data = Data(TArr(elem, 1))
+    val m = a.length
+    data.setArrayShape(m)
+    val bytes = data.arrayBytes
+    var i = 0
+    while (i < m) {
+      bytes.setBool(elem.dataWidth * i, a(i))
+      i += 1
+    }
+    data
+  }
+  def apply(a: Array[Int]): Data = {
+    val elem = TInt
+    val data = Data(TArr(elem, 1))
+    val m = a.length
+    data.setArrayShape(m)
+    val bytes = data.arrayBytes
+    var i = 0
+    while (i < m) {
+      bytes.setInt(elem.dataWidth * i, a(i))
+      i += 1
+    }
+    data
+  }
+  def apply(a: Array[Long]): Data = {
+    val elem = TUInt
+    val data = Data(TArr(elem, 1))
+    val m = a.length
+    data.setArrayShape(m)
+    val bytes = data.arrayBytes
+    var i = 0
+    while (i < m) {
+      bytes.setUInt(elem.dataWidth * i, a(i))
+      i += 1
+    }
+    data
+  }
```

Comment:
So, after some more investigation, the array type switch does work as expected, but the generic `arrayApply` function seems to take up to 50% longer on the test case in `DataBench` for `Array[Int]`. (Strangely, it performs comparably on `Array[Double]`; not sure what the difference is in these two cases.) I'm going to leave the separate functions for now.